tfhe/src/safe_deserialization.rs

```diff
@@ -0,0 +1,205 @@
+use crate::conformance::ParameterSetConformant;
+use crate::Named;
+use bincode::Options;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+// The `SERIALIZATION_VERSION` is serialized along objects serialized with `safe_serialize`.
+// This `SERIALIZATION_VERSION` should be changed on each release where any object serialization
+// details changes (this can happen when adding a field or reorderging fields of a struct).
+// When a object is deserialized using `safe_deserialize`, the deserialized version is checked
+// to be equal to SERIALIZATION_VERSION.
+// This helps prevent users from inadvertently deserializaing an object serialized in another
+// release.
+// When this happens, it also gives a clear version mismatch error rather than a generic
+// deserialization error or worse, a garbage object.
+const SERIALIZATION_VERSION: &str = "0.1";
+
+// `VERSION_LENGTH_LIMIT` is the maximum `SERIALIZATION_VERSION` size which `safe_deserialization`
+// is going to try to read (it returns an error if it's too big).
+// It helps prevent an attacker passing a very long `SERIALIZATION_VERSION` to exhaust memory.
+const VERSION_LENGTH_LIMIT: u64 = 100;
+
+const TYPE_NAME_LENGTH_LIMIT: u64 = 1000;
+
+/// Serializes an object into a [writer](std::io::Write).
+/// The result contains a version of the serialization and the name of the
+/// serialized type to provide checks on deserialization with [safe_deserialize].
+/// `serialized_size_limit` is the size limit (in number of byte) of the serialized object
+/// (excluding version and name serialization).
+pub fn safe_serialize<T: Serialize + Named>(
+    object: &T,
+    mut writer: impl std::io::Write,
+    serialized_size_limit: u64,
+) -> bincode::Result<()> {
+    let options = bincode::DefaultOptions::new()
+        .with_fixint_encoding()
+        .with_limit(0);
+
+    options
+        .with_limit(VERSION_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &SERIALIZATION_VERSION.to_owned())?;
+
+    options
+        .with_limit(TYPE_NAME_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &T::NAME.to_owned())?;
```

Comment:
Do we really need these .to_owned() ? passign the &'static str should work

Reply:
For deserialization, I think we have to use the `String` type
I'm not sure a serialized `&str` is guaranted to be deserializable as a `String` which is why I did it this way